src/Utilities/FlowAnalysis/FlowAnalysis/Analysis/TaintedDataAnalysis/TaintedDataAnalysis.TaintedDataOperationVisitor.cs

```diff
@@ -226,54 +224,65 @@ public override TaintedDataAbstractValue VisitObjectCreation(IObjectCreationOper
                     if (this.IsSanitizingMethod(method))
                     {
-                        return TaintedDataAbstractValue.NotTainted;
+                        result = TaintedDataAbstractValue.NotTainted;
                     }
                     else if (visitedInstance != null && this.IsSanitizingInstanceMethod(method))
                     {
-                        if (AnalysisEntityFactory.TryCreate(visitedInstance, out AnalysisEntity analysisEntity))
-                        {
-                            this.CurrentAnalysisData.SetAbstractValue(analysisEntity, TaintedDataAbstractValue.NotTainted);
-                        }
-
-                        return TaintedDataAbstractValue.NotTainted;
+                        result = TaintedDataAbstractValue.NotTainted;
+                        SetTaintedForEntity(visitedInstance, result);
                     }
                     else if (this.DataFlowAnalysisContext.SourceInfos.IsSourceMethod(
                         method,
-                        out evaluateWithPointsToAnalysis,
-                        out evaluateWithValueContentAnalysis))
+                        visitedArguments,
+                        (originalOperation as IInvocationOperation).Arguments.Select(o => GetPointsToAbstractValue(o.Value)).ToImmutableArray(),
+                        (originalOperation as IInvocationOperation).Arguments.Select(o => GetValueContentAbstractValue(o.Value)).ToImmutableArray(),
+                        out taintedTargets))
                     {
-                        ImmutableArray<IArgumentOperation> argumentOperation = (originalOperation as IInvocationOperation).Arguments;
-                        IEnumerable<PointsToAbstractValue> pointsToAnalysisResult = argumentOperation.Select(o => GetPointsToAbstractValue(o.Value));
-                        IEnumerable<ValueContentAbstractValue> valueContentAnalysisResult = argumentOperation.Select(o => GetValueContentAbstractValue(o.Value));
-                        if ((evaluateWithPointsToAnalysis != null && evaluateWithPointsToAnalysis.Any(o => o(pointsToAnalysisResult)))
-                            || (evaluateWithValueContentAnalysis != null && evaluateWithValueContentAnalysis.Any(o => o(pointsToAnalysisResult, valueContentAnalysisResult))))
+                        foreach (string taintedTarget in taintedTargets)
                         {
-                            return TaintedDataAbstractValue.CreateTainted(method, originalOperation.Syntax, this.OwningSymbol);
+                            if (taintedTarget != TaintedTargetValue.Return)
+                            {
+                                IArgumentOperation argumentOperation = visitedArguments.FirstOrDefault(o => o.Parameter.Name == taintedTarget);
+                                this.CacheAbstractValue(argumentOperation, TaintedDataAbstractValue.CreateTainted(argumentOperation.Parameter, argumentOperation.Syntax, method));
+                            }
```

Comment:
Just in case someone misconfigures tainted data sources, let's check if argumentOperation != null (and Debug.Fail() if it is null).

Reply:
Need an `if` statement so we fail gracefully if argumentOperation is null (don't want a NullReferenceException).